codec-http2/src/test/java/io/netty/handler/codec/http2/Http2CodecTest.java

```diff
@@ -0,0 +1,144 @@
+public class Http2CodecTest {
+
+    private EmbeddedChannel channel;
+
+    private Http2FrameWriter frameWriter;
+    private TestChannelInitializer testChannelInitializer;
+
+    @Before
+    public void setUp() {
+        frameWriter = spy(new VerifiableHttp2FrameWriter());
+        testChannelInitializer = new TestChannelInitializer();
+        Http2StreamChannelBootstrap bootstrap = new Http2StreamChannelBootstrap().handler(testChannelInitializer);
+        channel = new EmbeddedChannel();
+        channel.connect(new InetSocketAddress(0));
+        channel.pipeline().addLast(new Http2Codec(true, bootstrap, frameWriter));
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        channel.finishAndReleaseAll();
+    }
+
+    @Test
+    public void multipleOutboundStreams() {
+        Http2StreamChannelBootstrap b = new Http2StreamChannelBootstrap();
+        b.parentChannel(channel).handler(testChannelInitializer);
+
+        Channel channel1 = b.connect().channel();
+        assertTrue(channel1.isActive());
+        assertFalse(((AbstractHttp2StreamChannel) channel1).hasStreamId());
+        Channel channel2 = b.connect().channel();
+        assertTrue(channel2.isActive());
+        assertFalse(((AbstractHttp2StreamChannel) channel2).hasStreamId());
+
+        Http2Headers headers1 = new DefaultHttp2Headers();
+        Http2Headers headers2 = new DefaultHttp2Headers();
+        // Test that streams can be made active (headers sent) in different order than the corresponding channels
+        // have been created.
+        channel2.writeAndFlush(new DefaultHttp2HeadersFrame(headers2));
+        channel1.writeAndFlush(new DefaultHttp2HeadersFrame(headers1));
+
+        verify(frameWriter).writeHeaders(any(ChannelHandlerContext.class), eq(2), same(headers2), anyInt(), anyShort(),
+                                         eq(false), eq(0), eq(false), any(ChannelPromise.class));
+        verify(frameWriter).writeHeaders(any(ChannelHandlerContext.class), eq(4), same(headers1), anyInt(), anyShort(),
+                                         eq(false), eq(0), eq(false), any(ChannelPromise.class));
+
+        assertTrue(((AbstractHttp2StreamChannel) channel1).hasStreamId());
+        assertTrue(((AbstractHttp2StreamChannel) channel2).hasStreamId());
+
+        channel1.close();
+        channel2.close();
+    }
+
+    @Test
+    public void createOutboundStream() {
+        Http2StreamChannelBootstrap b = new Http2StreamChannelBootstrap();
+        ChannelFuture channelFuture = b.parentChannel(channel).handler(testChannelInitializer).connect();
+        assertTrue(channelFuture.isSuccess());
+        Channel childChannel = channelFuture.channel();
+        assertTrue(childChannel.isRegistered());
+        assertTrue(childChannel.isActive());
+
+        Http2Headers headers = new DefaultHttp2Headers();
+        childChannel.write(new DefaultHttp2HeadersFrame(headers));
+        ByteBuf data = Unpooled.buffer(100).writeZero(100);
```

Comment:
If I put a break point here and step through I get an exception on line 114 (the first `verify(frameWrite).writeHeaders(...)` call. It looks like something goes wrong with the mocking and the framewriter method is actually called with `ChannelHandlerContext == null` ... any ideas (I don't recall seeing this with other mocked objects)?

```
java.lang.NullPointerException
	at io.netty.handler.codec.http2.DefaultHttp2FrameWriter.writeHeadersInternal(DefaultHttp2FrameWriter.java:414)
	at io.netty.handler.codec.http2.DefaultHttp2FrameWriter.writeHeaders(DefaultHttp2FrameWriter.java:194)
	at io.netty.handler.codec.http2.Http2CodecTest.createOutboundStream(Http2CodecTest.java:114)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:117)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:262)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:84)
```